File: app/src/main/java/com/jakebarnby/filemanager/models/GoogleDriveSource.java
package com.jakebarnby.filemanager.models;

import android.Manifest;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.support.v4.app.Fragment;

import com.google.android.gms.auth.GoogleAuthException;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.client.util.ExponentialBackOff;
import com.google.api.services.drive.DriveScopes;
import com.google.api.services.drive.model.File;
import com.google.api.services.drive.model.FileList;
import com.jakebarnby.filemanager.managers.GoogleDriveFactory;
import com.jakebarnby.filemanager.models.files.GoogleDriveFile;
import com.jakebarnby.filemanager.models.files.SourceFile;
import com.jakebarnby.filemanager.util.Constants;
import com.jakebarnby.filemanager.util.TreeNode;
import com.jakebarnby.filemanager.util.Utils;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

import static com.jakebarnby.filemanager.util.Constants.RequestCodes.ACCOUNT_PICKER;

/**
 * Created by jakebarnby on 2/08/17.
 */

public class GoogleDriveSource extends Source {

    private static final String[] SCOPES = {DriveScopes.DRIVE};
    private GoogleAccountCredential mCredential;

    public GoogleDriveSource(String sourceName, SourceListener listener) {
        super(sourceName, listener);

    }

    @Override
    public void authenticateSource(Context context) {
        mSourceListener.onCheckPermissions(Manifest.permission.GET_ACCOUNTS, Constants.RequestCodes.ACCOUNTS_PERMISSIONS);
    }

    @Override
    public void loadSource(Context context) {
        if (!isFilesLoaded()) {
            if (!checkConnectionActive(context)) return;
            new GoogleDriveFileSystemLoader(mCredential).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, "root");
        }
    }

    public void setCredential(Context context) {
        mCredential = GoogleAccountCredential
                .usingOAuth2(context, Arrays.asList(SCOPES))
                .setBackOff(new ExponentialBackOff());
    }

    public void authGoogle(Fragment fragment) {
        getResultsFromApi(fragment);
    }

    /**
     * Create a google credential and try to call the API with it, do nothing if it fails
     */
    public void authGoogleSilent(Fragment fragment) {
        String accountName = fragment.getContext().getSharedPreferences(Constants.SharedPrefs.PREFS, Context.MODE_PRIVATE)
                .getString(Constants.SharedPrefs.GOOGLE_ACCOUNT_NAME, null);
        if (accountName != null) {
            mCredential.setSelectedAccountName(accountName);
            getResultsFromApi(fragment);
        }
    }

    /**
     * Attempt to call the API, after verifying that all the preconditions are satisfied
     */
    public void getResultsFromApi(Fragment fragment) {
        if (!isLoggedIn()) {
            if (!isGooglePlayServicesAvailable(fragment.getContext())) {
                acquireGooglePlayServices(fragment.getActivity());
            } else if (mCredential != null && mCredential.getSelectedAccountName() == null) {
                fragment.startActivityForResult(mCredential.newChooseAccountIntent(), ACCOUNT_PICKER);
            } else if (!Utils.isConnectionReady(fragment.getContext())) {
                mSourceListener.onNoConnection();
            } else {
                setLoggedIn(true);
                loadSource(fragment.getContext());
            }
        }
    }

    public void saveUserToken(Fragment fragment) {
        SharedPreferences prefs = fragment.getContext().getSharedPreferences(Constants.SharedPrefs.PREFS, Context.MODE_PRIVATE);
        try {
            prefs.edit().putString(Constants.SharedPrefs.GOOGLE_ACCESS_TOKEN, mCredential.getToken()).apply();
        } catch (IOException | GoogleAuthException e) {
            e.printStackTrace();
        }
        getResultsFromApi(fragment);
    }

    public void saveUserAccount(Fragment fragment, String accountName) {
        SharedPreferences prefs = fragment.getContext().getSharedPreferences(Constants.SharedPrefs.PREFS, Context.MODE_PRIVATE);
        prefs.edit().putString(Constants.SharedPrefs.GOOGLE_ACCOUNT_NAME, accountName).apply();
        mCredential.setSelectedAccountName(accountName);
        getResultsFromApi(fragment);
    }

    /**
     * An asynchronous task that handles the Drive API call.
     * Placing the API calls in their own task ensures the UI stays responsive.
     */
    private class GoogleDriveFileSystemLoader extends AsyncTask<String, Void, TreeNode<SourceFile>> {

        private TreeNode<SourceFile> rootFileTreeNode;
        private TreeNode<SourceFile> currentLevelNode;

        GoogleDriveFileSystemLoader(GoogleAccountCredential credential) {
            HttpTransport transport = new NetHttpTransport();
            JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
            GoogleDriveFactory.getInstance().setService(
                    new com.google.api.services.drive.Drive.Builder(transport, jsonFactory, credential)
                            .setApplicationName("File Manager Android")
                            .build());

        }

        @Override
        protected void onPreExecute() {
            mSourceListener.onLoadStarted();
        }

        @Override
        protected TreeNode<SourceFile> doInBackground(String... paths) {
            try {
                File rootFile = GoogleDriveFactory
                        .getInstance()
                        .getService()
                        .files()
                        .get(paths[0])
                        .setFields("name,id,mimeType,parents,size,hasThumbnail,thumbnailLink,iconLink,modifiedTime")
                        .execute();
                SourceFile rootSourceFile = new GoogleDriveFile(rootFile);
                rootFileTreeNode = new TreeNode<>(rootSourceFile);
                currentLevelNode = rootFileTreeNode;
                setCurrentDirectory(rootFileTreeNode);
                setQuotaInfo(GoogleDriveFactory.getInstance().getStorageStats());

                return parseDirectory(rootFile);
            } catch (Exception e) {
                //TODO: Error handling
                return null;
            }
        }

        private TreeNode<SourceFile> parseDirectory(File currentDirectory) throws IOException {
            FileList fileList = GoogleDriveFactory
                    .getInstance()
                    .getService()
                    .files()
                    .list()
                    .setQ(String.format("'%s' in parents", currentDirectory.getId()))
                    .setFields("files(name,id,mimeType,parents,size,hasThumbnail,thumbnailLink,iconLink,modifiedTime)")
                    .execute();
            List<File> files = fileList.getFiles();
            long dirSize = 0L;
            if (files != null) {
                for (File file : files) {
                    SourceFile sourceFile = new GoogleDriveFile(file);
                    if (sourceFile.isDirectory()) {
                        currentLevelNode.addChild(sourceFile);
                        currentLevelNode = currentLevelNode.getChildren().get(currentLevelNode.getChildren().size() - 1);
                        parseDirectory(file);
                        currentLevelNode.getParent().getData().addSize(currentLevelNode.getData().getSize());
                        currentLevelNode = currentLevelNode.getParent();
                    } else {
                        if (file.getSize() != null) {
                            dirSize += file.getSize();
                        }
                        currentLevelNode.addChild(sourceFile);
                    }
                }
                currentLevelNode.getData().addSize(dirSize);
            }
            return rootFileTreeNode;
        }

        @Override
        protected void onPostExecute(TreeNode<SourceFile> fileTree) {
            super.onPostExecute(fileTree);
            TreeNode.sortTree(fileTree, (node1, node2) -> {
                int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                if (result == 0) {
                    result = node1.getData().getName().toLowerCase().compareTo(node2.getData().getName().toLowerCase());
                }
                return result;
            });
            setFilesLoaded(true);
            mSourceListener.onLoadComplete(fileTree);
        }
    }

    /**
     * Check that Google Play services APK is installed and up to date.
     * @return True if Google Play Services is available and up to date on this device, false otherwise.
     */
    public boolean isGooglePlayServicesAvailable(Context context) {
        GoogleApiAvailability apiAvailability = GoogleApiAvailability.getInstance();
        final int connectionStatusCode = apiAvailability.isGooglePlayServicesAvailable(context);
        return connectionStatusCode == ConnectionResult.SUCCESS;
    }

    /**
     * Attempt to resolve a missing, out-of-date, invalid or disabled Google
     * Play Services installation via a user dialog, if possible.
     */
    public void acquireGooglePlayServices(Activity activity) {
        GoogleApiAvailability apiAvailability = GoogleApiAvailability.getInstance();
        final int connectionStatusCode =
                apiAvailability.isGooglePlayServicesAvailable(activity);
        if (apiAvailability.isUserResolvableError(connectionStatusCode)) {
            showGooglePlayServicesAvailabilityErrorDialog(activity, connectionStatusCode);
        }
    }

    /**
     * Display an error dialog showing that Google Play Services is missing
     * or out of date.
     * @param connectionStatusCode code describing the presence (or lack of)
     *     Google Play Services on this device.
     */
    public static void showGooglePlayServicesAvailabilityErrorDialog(Activity activity, final int connectionStatusCode) {
        GoogleApiAvailability apiAvailability = GoogleApiAvailability.getInstance();
        Dialog dialog = apiAvailability.getErrorDialog(
                activity,
                connectionStatusCode,
                Constants.RequestCodes.GOOGLE_PLAY_SERVICES);
        dialog.show();
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/models/sources/googledrive/GoogleDriveSource.kt
package com.jakebarnby.filemanager.models.sources.googledrive

import android.Manifest
import android.content.Context
import android.webkit.MimeTypeMap
import androidx.fragment.app.Fragment
import androidx.room.Entity
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.http.FileContent
import com.google.api.client.util.ExponentialBackOff
import com.google.api.services.drive.Drive
import com.google.api.services.drive.DriveScopes
import com.google.api.services.drive.model.FileList
import com.google.firebase.analytics.FirebaseAnalytics
import com.jakebarnby.filemanager.managers.PreferenceManager
import com.jakebarnby.filemanager.models.Source
import com.jakebarnby.filemanager.models.SourceConnectionType
import com.jakebarnby.filemanager.models.SourceType
import com.jakebarnby.filemanager.models.StorageInfo
import com.jakebarnby.filemanager.models.sources.googledrive.params.*
import com.jakebarnby.filemanager.util.Constants
import com.jakebarnby.filemanager.util.Constants.Prefs
import com.jakebarnby.filemanager.util.Constants.RequestCodes
import com.jakebarnby.filemanager.util.GooglePlay
import com.jakebarnby.filemanager.util.Logger
import com.jakebarnby.filemanager.util.Utils
import com.jakebarnby.filemanager.workers.GoogleDriveFileTreeWalker
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import javax.inject.Inject

/**
 * Created by jakebarnby on 2/08/17.
 */
@Entity
class GoogleDriveSource @Inject constructor(
    prefsManager: PreferenceManager
) : Source<
    GoogleDriveCreateFolderParams,
    GoogleDriveDownloadParams,
    GoogleDriveUploadParams,
    GoogleDriveRenameParams,
    GoogleDriveDeleteParams,
    GoogleDriveFileTreeWalker,
    com.google.api.services.drive.model.File,
    com.google.api.services.drive.model.File
    >(
    SourceConnectionType.REMOTE,
    SourceType.GOOGLE_DRIVE.id,
    prefsManager
) {

    companion object {
        private val SCOPES = listOf(DriveScopes.DRIVE)

        var client: Drive? = null
    }

    private var credential: GoogleAccountCredential? = null

    override val storageInfo: StorageInfo?
        get() {
            try {
                val quota = client
                    ?.about()
                    ?.get()
                    ?.setFields("storageQuota")
                    ?.execute()
                    ?.storageQuota
                return StorageInfo(
                    quota?.limit ?: 0,
                    quota?.usage ?: 0
                )
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return null
        }

    override suspend fun authenticate(context: Context) {
        //if (!checkConnectionActive(context)) return
        fetchCredential(context)
        if (prefsManager.hasSourceToken(sourceId)) {
            loadFiles<GoogleDriveFileTreeWalker>(context)
            return
        }
        listener.onCheckPermissions(
            Manifest.permission.GET_ACCOUNTS,
            RequestCodes.ACCOUNTS_PERMISSIONS
        )
    }

    override suspend fun logout(context: Context) {
        if (isLoggedIn) {
            prefsManager.savePref(Prefs.GOOGLE_TOKEN_KEY, null as String?)
            prefsManager.savePref(Prefs.GOOGLE_NAME_KEY, null as String?)

            isLoggedIn = false
            isFilesLoaded = false
            credential = null
            listener.onLogout()

            Logger.logFirebaseEvent(
                FirebaseAnalytics.getInstance(context),
                Constants.Analytics.EVENT_LOGOUT_GOOGLEDRIVE
            )
        }
    }

    private fun fetchCredential(context: Context) {
        credential = GoogleAccountCredential
            .usingOAuth2(context, SCOPES)
            .setBackOff(ExponentialBackOff())
    }

    fun authenticate(fragment: Fragment) {
        getResultsFromApi(fragment)
    }

    /**
     * Create a google credential and try to call the API with it, do nothing if it fails
     */
    fun authGoogleSilent(fragment: Fragment) {
        val accountName = prefsManager.getString(
            Prefs.GOOGLE_NAME_KEY,
            null
        )
        if (accountName != null) {
            fetchCredential(fragment.context!!)
            credential!!.selectedAccountName = accountName
            getResultsFromApi(fragment)
        }
    }

    /**
     * Attempt to call the API, after verifying that all the preconditions are satisfied
     */
    fun getResultsFromApi(fragment: Fragment) {
        if (isLoggedIn) {
            return
        }

        if (!GooglePlay.isGooglePlayServicesAvailable(fragment.context!!)) {
            GooglePlay.acquireGooglePlayServices(fragment.activity)
        } else if (credential != null && credential!!.selectedAccountName == null) {
            fragment.startActivityForResult(credential!!.newChooseAccountIntent(), RequestCodes.ACCOUNT_PICKER)
        } else if (!Utils.isConnectionReady(fragment.context!!)) {
            listener.onNoConnection()
        } else {
            isLoggedIn = true
            loadFiles<GoogleDriveFileTreeWalker>(fragment.context!!)
        }
    }

    fun saveUserToken(fragment: Fragment) {
        try {
            prefsManager.savePref(Prefs.GOOGLE_TOKEN_KEY, credential!!.token)
        } catch (e: Exception) {
            e.printStackTrace()
            //TODO: Log error
        }
        getResultsFromApi(fragment)
    }

    fun saveUserAccount(fragment: Fragment, accountName: String?) {
        prefsManager.savePref(Prefs.GOOGLE_NAME_KEY, accountName)
        credential!!.selectedAccountName = accountName
        getResultsFromApi(fragment)
    }

    fun getRootFile() =
        client
            ?.files()
            ?.get("root")
            ?.setFields("name,id,mimeType,parents,size,hasThumbnail,thumbnailLink,iconLink,modifiedTime")
            ?.execute()

    fun getFilesByParentId(id: String): FileList? =
        client
            ?.files()
            ?.list()
            ?.setQ(String.format("'%s' in parents", id))
            ?.setFields("files(name,id,mimeType,parents,size,hasThumbnail,thumbnailLink,iconLink,modifiedTime)")
            ?.execute()


    @Throws(IOException::class)
    override suspend fun download(
        params: GoogleDriveDownloadParams
    ): File = withContext(IO) {
        val file = File(params.destinationPath)
        FileOutputStream(file).use {
            client
                ?.files()
                ?.get(params.id)
                ?.executeMediaAndDownloadTo(it)
        }
        file
    }

    @Throws(IOException::class)
    override suspend fun upload(
        params: GoogleDriveUploadParams
    ): com.google.api.services.drive.model.File? {
        val file = File(params.filePath)

        val mimeTypeMap = MimeTypeMap.getSingleton()
        val fileMimeType = mimeTypeMap.getMimeTypeFromExtension(Utils.fileExt(file.absolutePath))

        val fileMetadata = com.google.api.services.drive.model.File().apply {
            parents = listOf(params.parentId)
            mimeType = fileMimeType
            hasThumbnail = true
            name = params.fileName
        }

        val googleFile = FileContent(fileMimeType, file)

        return withContext(IO) {
            client
                ?.files()
                ?.create(fileMetadata, googleFile)
                ?.setFields("name,id,mimeType,parents,size,hasThumbnail,thumbnailLink,iconLink,modifiedTime")
                ?.execute()
        }
    }

    @Throws(IOException::class)
    override suspend fun delete(params: GoogleDriveDeleteParams) {
        withContext(IO) {
            client
                ?.files()
                ?.delete(params.fileId)
                ?.execute()
        }
    }

    @Throws(IOException::class)
    override suspend fun createFolder(params: GoogleDriveCreateFolderParams): com.google.api.services.drive.model.File? {
        val fileMetadata = com.google.api.services.drive.model.File().apply {
            parents = listOf(params.parentId)
            name = params.folderName
            mimeType = Constants.Sources.GOOGLE_DRIVE_FOLDER_MIME
        }

        return withContext(IO) {
            client
                ?.files()
                ?.create(fileMetadata)
                ?.setFields("name,id,mimeType,parents,size,hasThumbnail,thumbnailLink,iconLink,modifiedTime")
                ?.execute()
        }
    }

    @Throws(IOException::class)
    override suspend fun rename(params: GoogleDriveRenameParams
    ): com.google.api.services.drive.model.File? {
        val file = com.google.api.services.drive.model.File().apply {
            name = params.newName
        }

        return withContext(IO) {
            client
                ?.files()
                ?.update(params.fileId, file)
                ?.setFields("name")
                ?.execute()
        }
    }
}